File: app/models/Game.scala
package models

import models.TileOwner._
import TileOwner._
import com.mongodb.casbah.Imports._
import com.mongodb.casbah.MongoClientURI
import play.api.Play

object PlayerTurn extends Enumeration {
	type PlayerTurn = Value
	val PlayerOne = Value("1")
	val PlayerTwo = Value("2")
	
	def other(turn: PlayerTurn): PlayerTurn =
	  if (turn == PlayerOne)
	    PlayerTwo
	  else
	    PlayerOne
}

import PlayerTurn._

case class Game(id: String, tiles: List[Tile], playerOne: String, 
    playerTwo: Option[String], score: (Int, Int), 
    turn: PlayerTurn, words: List[Word], size: Int)

case class Word(word: String, turn: PlayerTurn)
    
object Game {
  
  val DEFAULT_SIZE = 5

  val uri = Play.current.configuration.getString("mongodb.uri").get
  val mongoClient =  MongoClient(MongoClientURI(uri))
  val mongoColl = mongoClient("letterpress")("game")
  val words = scala.io.Source.fromFile("conf/wordlist/fi.txt", "UTF-8").getLines.toVector

  def create(game: Game) { 
    val gameObj = MongoDBObject(
    	"id" -> game.id,
    	"tiles" -> serializeTiles(game.tiles),
    	"playerOne" -> game.playerOne,
    	"playerTwo" -> game.playerTwo,
    	"turn" -> game.turn.toString,
    	"words" -> List(),
    	"size" -> game.size
    )
    mongoColl += gameObj
  }
  
  def getCurrentTurn(game: Game) = 
    if (game.turn == PlayerTurn.PlayerOne)
      Some(game.playerOne)
    else
      game.playerTwo
  
  def updateTiles(id: String, turn: PlayerTurn, tiles: List[Tile]) {
    mongoColl.update(MongoDBObject("id" -> id), 
        $set(Seq("turn" -> turn.toString,
            "tiles" -> serializeTiles(tiles))))
  }
  
  def fetch(id: String): Option[Game] = {
    mongoColl.findOne(MongoDBObject("id" -> id)).map(
      game => {
    	  val tiles = deserializeTiles(game.getAsOrElse[String]("tiles", ""))
    	  Game(game.getAsOrElse[String]("id", null),
            tiles,
            game.getAsOrElse[String]("playerOne", ""),
            game.getAs[String]("playerTwo"),
            score(tiles),
            PlayerTurn.withName(game.getAsOrElse[String]("turn", PlayerTurn.PlayerOne.toString)),
            deserializeWords(game.as[MongoDBList]("words").toList collect { case s: String => s }),
            game.getAsOrElse[Int]("size", DEFAULT_SIZE)
          )
      }
    ) 
  }

  def serializeWord(word: Word): String =
    word.turn.toString + word.word
      
  def deserializeWords(words: List[String]): List[Word] = {
    def deserializeWord(word: String): Word =
      Word(word.substring(1), PlayerTurn.withName(word.substring(0, 1)))
 
    if (words.isEmpty)
      Nil
    else
      deserializeWord(words.head) :: deserializeWords(words.tail)
  }
  
  def join(id: String, name: String) {
    mongoColl.update(MongoDBObject("id" -> id), 
        $set(Seq("playerTwo" -> name)))
  }

  def turn2Owner(turn: PlayerTurn): TileOwner =
    if (turn == PlayerTurn.PlayerOne)
      TileOwner.PlayerOne
    else
      TileOwner.PlayerTwo
  
  def submit(game: Game, word: String, tiles: String, currentPlayer: Option[String]): String = {
    if (!tilesMatch(game, word, tiles))
      "Submitted tiles and word don't match."
    else if (currentPlayer.isEmpty || !currentPlayer.equals(getCurrentTurn(game)))
      "It's not your turn"
    else if (game.playerTwo.isEmpty && game.turn == PlayerTurn.PlayerTwo)
      "Player two name has not joined and can't play yet."
    else if (ended(game.tiles))
      "The game has ended"
    else if (game.words.count(_.word.startsWith(word)) > 0)
      word + " has already been played."
    else if (words.contains(word.toLowerCase())) {
      saveSubmit(word, game, tiles)
      "OK"
    } else
      word + " is not a valid word."
  }
      
  def saveSubmit(word: String, game: Game, tiles: String) {
	// save new word
    mongoColl.update(MongoDBObject("id" -> game.id),
        $push(Seq("words" -> serializeWord(Word(word, game.turn)))))
    
    val newTiles = normalize(Tile.selectWord(stringToTileIds(tiles), game.tiles, game.turn), game.size)
    updateTiles(game.id, other(game.turn), newTiles)
  }
  
  def stringToTileIds(tilesString: String): List[Int] = {
    tilesString.split(",").toList.map(s => s.toInt)    
  }
  
  def tilesMatch(game: Game, word: String, tilesString: String): Boolean = {
    val tiles = stringToTileIds(tilesString)
    
    // check for duplicate ids
    if (tiles.distinct != tiles)
      false
    else {
      val tileWord = tiles.map { id => game.tiles(id).letter }.mkString
      word == tileWord
    }
  }
  
  def pass(game: Game) {
    mongoColl.update(MongoDBObject("id" -> game.id),
        $set(Seq("turn" -> PlayerTurn.other(game.turn).toString)))
  }
  
  def serializeTiles(tiles: List[Tile]) =
   (for (t <- tiles) yield (t.letter + t.owner.toString)).mkString
  
   def deserializeTiles(tiles: String): List[Tile] = {
    def deserializeTiles0(i: Int, tiles: String): List[Tile] =
      if (tiles.length() < 2)
        Nil
      else {
        val c = tiles.charAt(0)
        val owner = TileOwner.withName(tiles.substring(1, 2))	
        Tile(c, i, owner) :: deserializeTiles0(i + 1, tiles.substring(2))
    }
    deserializeTiles0(0, tiles)
  }
  
  def score(tiles: List[Tile]): (Int, Int) = 
    (tiles.filter(t => locked(t.owner) == TileOwner.PlayerOneLocked).length,
    tiles.filter(t => locked(t.owner) == TileOwner.PlayerTwoLocked).length)
    
  def ended(tiles: List[Tile]): Boolean =
    tiles.count(_.owner == Neither) == 0

  def winner(game: Game): Option[String] =
    if (ended(game.tiles)) {
      val s = score(game.tiles)
      if (s._1 > s._2)
        Some(game.playerOne)
      else
        game.playerTwo
    } else
      None
      
   // *** Handle List of Tiles ***
  def normalize(tiles: List[Tile], gameSize: Int): List[Tile] = { 
    tiles map { t => normalize0(neighbors(tiles, t.id, gameSize), t) }
  }    
    
  def normalize0(neighbors: List[Tile], tile: Tile): Tile =
    if (tile.owner == Neither)
      tile
    else {
      val equalNeighbors = neighbors.map { t => locked(t.owner) }.
        count {_ == locked(tile.owner)}
      if (equalNeighbors == neighbors.length)
        Tile(tile.letter, tile.id, locked(tile.owner))
      else 
        Tile(tile.letter, tile.id, unlocked(tile.owner))
    }
    
  def neighbors(tiles: List[Tile], id: Int, gameSize: Int): List[Tile] = {
    val ROWS = gameSize
    val COLS = gameSize
    val col = id % ROWS
    val row = id / COLS
    List((col-1, row), (col+1, row), (col, row-1), (col, row+1)).
    	filter { c => c._1 >= 0 && c._2 >= 0 && c._1 < COLS && c._2 < ROWS }.
    	map { c => tiles(c._2 * COLS + c._1) }
  }
   
}

File: src/clj/letterpress/game.clj
(ns letterpress.game
  (:require [letterpress.db.postgresql :as db]
            [clojure.java.io :as io]
            [letterpress.tile :as tile]))

(def words
  (-> "fi.txt"
      io/resource
      slurp
      clojure.string/split-lines))

(defn- generate-id []
  (str (. java.util.UUID randomUUID)))

(defn create-game
  "Creates a new game, saves it in the database and
returns the game object."
  [player-name size]
  (let [size-num (clojure.edn/read-string size)
        game {:_id (generate-id)
              :tiles (tile/generate-tiles size-num)
              :player-one player-name
              :score [0 0]
              :turn :player-one
              :played-words []
              :size size-num}]
    (db/save-game game)
    game))

(defn list-games [player-name]
  (db/list-games player-name))

(defn get-game
  [id]
  (db/get-game id))

(defn join-game
  [id player-name]
  (let [result (db/join-game id player-name)]
    (when (= 1 result)
      (db/get-game id))))

(defn- is-turn?
  [game player-name]
  (-> game :turn keyword game (= player-name)))
      
(defn- turn-flip
  [turn]
  (if (= turn "player-one")
      "player-two"
      "player-one"))
  
(defn pass-game
  [id player-name]
  (let [game (db/get-game id)]
    (when (is-turn? game player-name)
      (db/update-game id
        (conj game
          {:turn (turn-flip (:turn game))}))
      (db/get-game id))))

(defn- game-contains-tiles?
  [game-tiles tiles]
  (->> tiles
       (map (fn [x] (some #(= x %) game-tiles)))
       (every? identity)))

(defn- is-played-word?
  [word played-words]
  (some #(.startsWith (:word %) word) played-words))

(defn- valid-submit?
  [word game tiles player-name]
  "Are `word` and `tiles` a valid submission for game state `game` and `player-name`?"
  (and (some (partial = word) words)
       (game-contains-tiles? (:tiles game) tiles)
       (is-turn? game player-name)
       (not (is-played-word? word (:played-words game)))))

(defn- new-score-fn
  [turn]
  (fn [old-score {owner :owner}]
    (if (or (= owner turn)
            (= owner "player-one-locked")
            (= owner "player-two-locked"))
      old-score
      (if (= owner (turn-flip turn)) ; other player owned the tile
        (if (= turn "player-one")
          [(-> old-score first inc) (-> old-score second dec)]
          [(-> old-score first dec) (-> old-score second inc)])
        (if (= turn "player-one") ; no one owned the tile
          [(-> old-score first inc) (-> old-score second)]
          [(-> old-score first) (-> old-score second inc)])))))

(defn- create-new-game-state
  [word game tiles]
  (let [new-tiles (tile/update-tiles game tiles)
        new-score (reduce (new-score-fn (:turn game)) (:score game) tiles)
        new-played-words (conj (:played-words game) {:word word :turn (:turn game)})
        tile-count (count new-tiles)
        score-sum (apply + new-score)
        winner (when (= score-sum tile-count)
                 {:winner (if (apply > new-score)
                            (:player-one game)
                            (:player-two game))})]
    (merge
      game
      {:tiles new-tiles
       :score new-score
       :turn (turn-flip (:turn game))
       :played-words new-played-words}
      winner)))

(defn submit-word
  [id player-name tiles]
  (let [word (->> tiles (map :letter) (reduce str) clojure.string/lower-case)
        game (db/get-game id)]
    (when (valid-submit? word game tiles player-name)
      (db/update-game
        id
        (create-new-game-state word game tiles))
      (db/get-game id))))

(defn- get-player
  [player-name game]
  (if (= player-name (:player-one game))
    "player-one"
    (when (= player-name (:player-two game))
      "player-two")))

(defn validate-word
  [id player-name tiles]
  (let [word (->> tiles (map :letter) (reduce str) clojure.string/lower-case)
        game (db/get-game id)
        player (get-player player-name game)]
    (if (nil? player)
      {:valid-word false}
      {:new-score (reduce (new-score-fn player) (:score game) tiles)
       :valid-word (valid-submit? word game tiles player-name)})))
